### PR TITLE
[new-parser] Merge dev-new-parser to main

### PR DESCRIPTION
**Pull request by tkobayas**
**Issue**: 

* https://github.com/apache/incubator-kie-drools/issues/5988

Merging `dev-new-parser` branch to `main` branch.

- Merging has to be done **after** ~~Drools 10.0 release~~ 10.0.x is branched out.
- ~~Until the Drools 10.0 release~~ Until 10.0.x is branched out, mark the PR [DO-NOT-MERGE] and make it **draft**.
- This PR makes the new parser **not** default parser. So it shouldn't affect the default behavior. 
- Make sure to use **"Rebase and merge"** (not "Squash and merge") in the Github UI to keep commit history.
- After Drools 10 release, discuss the future plan in dev at kie.apache ML.

### CODE REVIEW

**Comment by tkobayas** [comment]
[DRAFT] I'll ask for review **after** ~~Drools 10.0 is released~~ 10.0.x is branched out.

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/DrlParser.java [hunk 1]
```diff
@@ -49,12 +53,18 @@ public class DrlParser {
...
     private final List<DroolsError> results               = new ArrayList<>();
     private List<DroolsSentence>    editorSentences       = null;
     private Location                location              = new Location( Location.LOCATION_UNKNOWN );
     private DRLLexer                lexer                 = null;
     private Resource                resource              = null;
 
+    public static final String ANTLR4_PARSER_ENABLED_PROPERTY = "drools.drl.antlr4.parser.enabled";
+
+    // temporarily removed 'final' for testing purposes. This should be final when the feature gets stable
+    public static boolean ANTLR4_PARSER_ENABLED = Boolean.parseBoolean(System.getProperty(ANTLR4_PARSER_ENABLED_PROPERTY, "false")); // default is false
```
Comment (tkobayas) [inline drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/DrlParser.java:66]: The new parser is disabled by default. So it shouldn't affect the default behavior. Also unit tests (except `drools-drl-parser-tests`) run with the old parser.

https://github.com/apache/incubator-kie-drools/pull/5990 is for the new parser CI check.

**Comment by tkobayas** [comment]
Reminder:
Make sure to use **"Rebase and merge"** (not "Squash and merge") in the Github UI to keep commit history.

**Comment by tkobayas** [comment]
Hi @mariofusco @gitgabrio @baldimir @pibizza @lucamolteni ,

This is a PR to merge antlr4 based new parser. Note that this is not a default parser. So it shouldn't affect the default behavior. We would discuss the future plan in dev at kie.apache ML after 10.0.0 release.

This new parser pass all unit tests : Confirmed in https://github.com/apache/incubator-kie-drools/pull/5990

Please review, thanks!

**Review by mariofusco — APPROVED** [review]

**Review by lucamolteni — APPROVED** [review]

**Comment by tkobayas** [comment]
Rebased to resolve conflict with https://github.com/apache/incubator-kie-drools/pull/6115

**Review by pibizza — APPROVED** [review]
LGTM

**Pull request merged**